man/simplevisor.1
.\" Automatically generated by Pod::Man 2.27 (Pod::Simple 3.28)
.\"
.\" Standard preamble:
.\" ========================================================================
.de Sp \" Vertical space (when we can't use .PP)
.if t .sp .5v
.if n .sp
..
.de Vb \" Begin verbatim text
.ft CW
.nf
.ne \\$1
..
.de Ve \" End verbatim text
.ft R
.fi
..
.\" Set up some character translations and predefined strings.  \*(-- will
.\" give an unbreakable dash, \*(PI will give pi, \*(L" will give a left
.\" double quote, and \*(R" will give a right double quote.  \*(C+ will
.\" give a nicer C++.  Capital omega is used to do unbreakable dashes and
.\" therefore won't be available.  \*(C` and \*(C' expand to `' in nroff,
.\" nothing in troff, for use with C<>.
.tr \(*W-
.ds C+ C\v'-.1v'\h'-1p'\s-2+\h'-1p'+\s0\v'.1v'\h'-1p'
.ie n \{\
.    ds -- \(*W-
.    ds PI pi
.    if (\n(.H=4u)&(1m=24u) .ds -- \(*W\h'-12u'\(*W\h'-12u'-\" diablo 10 pitch
.    if (\n(.H=4u)&(1m=20u) .ds -- \(*W\h'-12u'\(*W\h'-8u'-\"  diablo 12 pitch
.    ds L" ""
.    ds R" ""
.    ds C` ""
.    ds C' ""
'br\}
.el\{\
.    ds -- \|\(em\|
.    ds PI \(*p
.    ds L" ``
.    ds R" ''
.    ds C`
.    ds C'
'br\}
.\"
.\" Escape single quotes in literal strings from groff's Unicode transform.
.ie \n(.g .ds Aq \(aq
.el       .ds Aq '
.\"
.\" If the F register is turned on, we'll generate index entries on stderr for
.\" titles (.TH), headers (.SH), subsections (.SS), items (.Ip), and index
.\" entries marked with X<> in POD.  Of course, you'll have to process the
.\" output yourself in some meaningful fashion.
.\"
.\" Avoid warning from groff about undefined register 'F'.
.de IX
..
.nr rF 0
.if \n(.g .if rF .nr rF 1
.if (\n(rF:(\n(.g==0)) \{
.    if \nF \{
.        de IX
.        tm Index:\\$1\t\\n%\t"\\$2"
..
.        if !\nF==2 \{
.            nr % 0
.            nr F 2
.        \}
.    \}
.\}
.rr rF
.\"
.\" Accent mark definitions (@(#)ms.acc 1.5 88/02/08 SMI; from UCB 4.2).
.\" Fear.  Run.  Save yourself.  No user-serviceable parts.
.    \" fudge factors for nroff and troff
.if n \{\
.    ds #H 0
.    ds #V .8m
.    ds #F .3m
.    ds #[ \f1
.    ds #] \fP
.\}
.if t \{\
.    ds #H ((1u-(\\\\n(.fu%2u))*.13m)
.    ds #V .6m
.    ds #F 0
.    ds #[ \&
.    ds #] \&
.\}
.    \" simple accents for nroff and troff
.if n \{\
.    ds ' \&
.    ds ` \&
.    ds ^ \&
.    ds , \&
.    ds ~ ~
.    ds /
.\}
.if t \{\
.    ds ' \\k:\h'-(\\n(.wu*8/10-\*(#H)'\'\h"|\\n:u"
.    ds ` \\k:\h'-(\\n(.wu*8/10-\*(#H)'\`\h'|\\n:u'
.    ds ^ \\k:\h'-(\\n(.wu*10/11-\*(#H)'^\h'|\\n:u'
.    ds , \\k:\h'-(\\n(.wu*8/10)',\h'|\\n:u'
.    ds ~ \\k:\h'-(\\n(.wu-\*(#H-.1m)'~\h'|\\n:u'
.    ds / \\k:\h'-(\\n(.wu*8/10-\*(#H)'\z\(sl\h'|\\n:u'
.\}
.    \" troff and (daisy-wheel) nroff accents
.ds : \\k:\h'-(\\n(.wu*8/10-\*(#H+.1m+\*(#F)'\v'-\*(#V'\z.\h'.2m+\*(#F'.\h'|\\n:u'\v'\*(#V'
.ds 8 \h'\*(#H'\(*b\h'-\*(#H'
.ds o \\k:\h'-(\\n(.wu+\w'\(de'u-\*(#H)/2u'\v'-.3n'\*(#[\z\(de\v'.3n'\h'|\\n:u'\*(#]
.ds d- \h'\*(#H'\(pd\h'-\w'~'u'\v'-.25m'\f2\(hy\fP\v'.25m'\h'-\*(#H'
.ds D- D\\k:\h'-\w'D'u'\v'-.11m'\z\(hy\v'.11m'\h'|\\n:u'
.ds th \*(#[\v'.3m'\s+1I\s-1\v'-.3m'\h'-(\w'I'u*2/3)'\s-1o\s+1\*(#]
.ds Th \*(#[\s+2I\s-2\h'-\w'I'u*3/5'\v'-.3m'o\v'.3m'\*(#]
.ds ae a\h'-(\w'a'u*4/10)'e
.ds Ae A\h'-(\w'A'u*4/10)'E
.    \" corrections for vroff
.if v .ds ~ \\k:\h'-(\\n(.wu*9/10-\*(#H)'\s-2\u~\d\s+2\h'|\\n:u'
.if v .ds ^ \\k:\h'-(\\n(.wu*10/11-\*(#H)'\v'-.4m'^\v'.4m'\h'|\\n:u'
.    \" for low resolution devices (crt and lpr)
.if \n(.H>23 .if \n(.V>19 \
\{\
.    ds : e
.    ds 8 ss
.    ds o a
.    ds d- d\h'-1'\(ga
.    ds D- D\h'-1'\(hy
.    ds th \o'bp'
.    ds Th \o'LP'
.    ds ae ae
.    ds Ae AE
.\}
.rm #[ #] #H #V #F C
.\" ========================================================================
.\"
.IX Title "SIMPLEVISOR 1"
.TH SIMPLEVISOR 1 "2019-08-23" "" "simplevisor man page"
.\" For nroff, turn off justification.  Always turn off hyphenation; it makes
.\" way too many mistakes in technical documents.
.if n .ad l
.nh
.SH "NAME"
simplevisor \- simple daemons supervisor
.SH "SYNOPSIS"
.IX Header "SYNOPSIS"
\&\fBsimplevisor\fR
[\-\-conf \s-1CONF\s0][\-\-conftype \s-1CONFTYPE\s0][\-\-daemon][\-\-interval \s-1INTERVAL\s0][\-h][\-\-log \s-1LOG\s0][\-\-logfile \s-1LOGFILE\s0][\-\-loglevel \s-1LOGLEVEL\s0][\-\-logname \s-1LOGNAME\s0][\-p \s-1PIDFILE\s0][\-\-store \s-1STORE\s0][\-\-version]
command [path]
.SH "DESCRIPTION"
.IX Header "DESCRIPTION"
Simplevisor is a simple daemons supervisor, it is inspired
by Erlang \s-1OTP\s0 and it can supervise hierarchies of services.
.PP
\&\s-1COMMANDS\s0
.PP
If a path is given or only one service entry is given:
.PP
for a given X command
    run the service X command where service is the only entry provided
    or the entry identified by its path
.PP
If a path is given and the root entry is a supervisor:
.PP
restart_child
    tell a running simplevisor process to restart the child identified
    by the given path; it is different from the restart command as
    described above because, this way, we are sure that the running
    simplevisor will not attempt to check/start/stop the child while
    we restart it
.PP
If a path is not given and the root entry is a supervisor:
.PP
start
    start the simplevisor process which start the supervision.
    It can be used with \-\-daemon if you want it as daemon
.PP
stop
    stop the simplevisor process and all its children, if running
.PP
status
    return the status of the simplevisor process
.PP
check
    return the comparison between the expected state and the actual state.
    0 \-> everything is fine
    1 \-> warning, not expected
.PP
single
    execute one cycle of supervision and exit.
    Useful to be run in a cron script
.PP
wake_up
    tell a running simplevisor process to wake up and supervise
.PP
stop_supervisor
    only stop the simplevisor process but not the children
.PP
stop_children
    only stop the children but not the simplevisor process
.PP
check_configuration
    only check the configuration file
.PP
pod
    generate pod format help to be used by pod2man to generate man page
.PP
rst
    generate rst format help to be used in the web doc
.PP
help
    same as \-h/\-\-help, print help page
.SH "OPTIONS"
.IX Header "OPTIONS"
\&\fBpositional arguments:\fR
.Sp
.RS 4
\&\fBcommand\fR check, check_configuration, help, pod, restart, restart_child, rst, single, start, status, stop, stop_children, stop_supervisor, wake_up
.Sp
\&\fBpath\fR path to a service, subset of commands available: start, stop, status, check, restart
.RE
.PP
\&\fBoptional arguments:\fR
.Sp
.RS 4
\&\fB\-\-conf \s-1CONF\s0\fR configuration file
.Sp
\&\fB\-\-conftype \s-1CONFTYPE\s0\fR configuration file type (default: apache)
.Sp
\&\fB\-\-daemon\fR daemonize, \s-1ONLY\s0 with start
.Sp
\&\fB\-\-interval \s-1INTERVAL\s0\fR interval to wait between supervision cycles (default: 60)
.Sp
\&\fB\-h, \-\-help\fR print the help page
.Sp
\&\fB\-\-log \s-1LOG\s0\fR available: null, file, syslog, stdout (default: stdout)
.Sp
\&\fB\-\-logfile \s-1LOGFILE\s0\fR log file, \s-1ONLY\s0 for file
.Sp
\&\fB\-\-loglevel \s-1LOGLEVEL\s0\fR log level (default: warning)
.Sp
\&\fB\-\-logname \s-1LOGNAME\s0\fR log name (default: simplevisor)
.Sp
\&\fB\-p, \-\-pidfile \s-1PIDFILE\s0\fR the pidfile
.Sp
\&\fB\-\-store \s-1STORE\s0\fR file where to store the state, it is not mandatory, however recommended to store the simplevisor nodes status between restarts
.Sp
\&\fB\-\-version\fR print the program version
.RE
.SH "EXAMPLES"
.IX Header "EXAMPLES"
Create and edit the main configuration file:
.PP
.Vb 1
\&    ## look for simplevisor.conf.example in the examples.
.Ve
.PP
Run it:
.PP
.Vb 1
\&    simplevisor \-\-conf /path/to/simplevisor.conf start
.Ve
.PP
to run it in daemon mode:
.PP
.Vb 1
\&    simplevisor \-\-conf /path/to/simplevisor.conf \-\-daemon start
.Ve
.PP
For other commands:
.PP
.Vb 1
\&    simplevisor \-\-help
.Ve
.PP
Given the example configuration, to start the httpd service:
.PP
.Vb 1
\&    simplevisor \-\-conf /path/to/simplevisor.conf start svisor1/httpd
.Ve
.SH "AUTHOR"
.IX Header "AUTHOR"
Massimo Paladin <massimo.paladin@gmail.com>
.PP
Copyright (C) \s-1CERN 2013\-2021\s0


CHANGES
Change log for python-simplevisor:

1.3	Wed Nov  3 2021
	* [lc] packaging and code cleanup
	* [lc] tested with Python 3.11

1.2	Mon Jun 27 2016
	* [lc] added support for JSON configuration files
	* [lc] fixed bug in restart_child

1.1	Wed Oct  1 2014
        * [lc] now using runuser instead of sudo to change user

1.0	Tue Aug  5 2014
	* [lc] renamed configuration_check to check_configuration
	* [lc] added a wake_up command
	* [lc] added a restart_child command

0.9	Thu Feb 27 2014
	* [mp] allow including same file multiple times in the configuration
	* [lc] documentation cleanup
	* [lc] allow including multiple files via globbing
	* [lc] always use the tempfile module for temporary directories
	* [lc] added a logname option to better control how messages are logged

0.8	Fri Apr 26 2013
	* [mp] strategies reviewed and reimplemented
	* [mp] changing logging to python logging

0.7.1	Tue Jan 29 2013
	* [mp] handling correctly file redirection when daemonizing

0.7	Mon Jan 28 2013
	* [mp] better handling of json module import
	* [mp] fixing bug regarding function call with keywords in python < 2.6.5
	* [mp] simplevisor-loop updated
	* [mp] documentation improved
	* [mp] log handling improved
	* [mp] bug fixing
	* [mp] variable interpolation in configuration file
	* [mp] failure if adjustments > value in a window number of cycles

0.6	Thu Jan 17 2013
	* [mp] bug fixes and improvements
	* [mp] service daemonization functionality embedded

0.5	Fri Sep 14 2012
	* [mp] various cleaning and improvements

0.4	Thu Jul 5 2012
	* [mp] simplevisor-new-instance script added to the examples
	* [mp] simplevisor-loop added, new script to daemonize foreground programs

0.3	Wed Jun 27 2012
	* [mp] improvements

0.2	Wed Mar 14 2012
	* [mp] various improvements

0.1	Mon Mar 11 2012
	* [mp] first release

[lc] = Lionel Cons <lionel.cons@cern.ch>
[mp] = Massimo Paladin <Massimo.Paladin@gmail.com>


simplevisor/Simplevisor.py
"""
Simplevisor module.


Copyright (C) CERN 2013-2021
"""
import logging
import os
import signal
import sys
import time

import mtb.log as log
import mtb.pid
from mtb.pid import \
    pid_check, pid_quit, pid_read, pid_remove, \
    pid_status, pid_touch, pid_write
from mtb.prnt import print_nested_list
from mtb.proc import daemonize
from mtb.modules import json
from mtb.validation import get_int_or_die

from simplevisor.errors import SimplevisorError
from simplevisor.supervisor import Supervisor
from simplevisor import service, supervisor


QUICK_COMMANDS = ["status", "stop", "stop_supervisor", "stop_children",
                  "wake_up"]
SERVICE_COMMANDS = ["start", "stop", "status", "check", "restart"]
OTHER_COMMANDS = ["single", "check_configuration", "restart_child"]

DEFAULT_INTERVAL = 60


class Simplevisor(object):
    """ Simplevisor class. """
    prog = "simplevisor"

    def __init__(self, config=None, child_config=None):
        """ Initialize Simplevisor. """
        if config is None:
            config = dict()
        elif type(config) != dict:
            raise SimplevisorError("Simplevisor config is not a dictionary")
        if "logname" not in config:
            config["logname"] = self.__class__.__name__
        self.logger = logging.getLogger(config["logname"])
        mtb.pid.LOGGER = self.logger
        self._config = config
        self._status_file = self._config.get("store")
        self._running = False
        if child_config is None:
            self._child = None
        else:
            if "logname" not in child_config:
                child_config["logname"] = config["logname"]
            self._child = supervisor.new_child(child_config)

    def get_child(self, path=""):
        """ Return child by its path. """
        path_list = path.split("/")
        first = path_list.pop(0)
        if not first:
            return self._child
        if self._child.name == first:
            if not path_list:
                return self._child
            try:
                return self._child.get_child(path_list)
            except StandardError:
                pass
        raise ValueError("given path is invalid: %s" % path)

    @log.print_only_exception_error()
    def work(self):
        """ Controller. """
        command = self._config.get("command", "status")
        path = self._config.get("path", None)
        if path is not None and command == "restart_child":
            self.send_action(command + " " + path)
            return
        if path is None and command in QUICK_COMMANDS:
            getattr(self, command)()
            return
        if self._child is None:
            raise SimplevisorError("no entry found")
        if path is None and isinstance(self._child, Supervisor):
            if command != "check_configuration":
                self.load_status()
            getattr(self, command)()
            return
        if command not in SERVICE_COMMANDS:
            raise ValueError("command must be one of: %s" %
                             ", ".join(SERVICE_COMMANDS))
        if path is None:
            target = self._child
        else:
            target = self.get_child(path)
        if command == "check":
            self.check(target)
            return
        self.logger.debug("calling %s.%s", target.name, command)
        (return_code, out, err) = getattr(target, command)()
        if len(out.strip()) > 0:
            print("stdout: %s" % (out.strip(), ))
        if len(err.strip()) > 0:
            print("stderr: %s" % (err.strip(), ))
        sys.exit(return_code)

    def check_configuration(self):
        """ Only check the configuration. """
        # so far so good
        print("the configuration file is valid")
        sys.exit(0)

    def on_signal(self, signum, _):
        """ Handle signals. """
        if signum == signal.SIGINT:
            self.logger.info("caught SIGINT")
            self._running = False
        elif signum == signal.SIGTERM:
            self.logger.info("caught SIGTERM")
            self._running = False
        elif signum == signal.SIGHUP:
            self.logger.info("caught SIGHUP, ignoring it")

    def start(self):
        """ Do start action. """
        signal.signal(signal.SIGINT, self.on_signal)
        signal.signal(signal.SIGTERM, self.on_signal)
        signal.signal(signal.SIGHUP, self.on_signal)
        self.pre_run()
        if self._config.get("daemon"):
            daemonize()
            run_function = log.log_exceptions(
                logger_name=self.prog,
                re_raise=False)(self.run)
        else:
            run_function = log.log_exceptions(
                logger_name=self.prog,
                re_raise=True)(self.run)
        if self._config.get("pidfile"):
            pid_write(self._config["pidfile"], os.getpid(), excl=True)
        try:
            run_function()
        except Exception:
            self.save_status()
            if self._config.get("pidfile"):
                pid_remove(self._config.get("pidfile"))
            raise sys.exc_info()[1]
        if self._config.get("pidfile"):
            pid_remove(self._config.get("pidfile"))

    def single(self):
        """ Do single action. """
        if self._config.get("pidfile"):
            pid_write(self._config["pidfile"], os.getpid(), excl=True)
        self.pre_run()
        self.supervise()
        self.save_status()
        if self._config.get("pidfile"):
            pid_remove(self._config.get("pidfile"))

    def stop(self, action="quit"):
        """ Quit the process. """
        if not self._config.get("pidfile"):
            raise SimplevisorError("%s requires a pidfile" % action)
        pid = pid_read(self._config["pidfile"])
        timeout = 10
        if pid and timeout is not None:
            print("%s (pid %d) is being told to %s..." %
                  (self.prog, pid, action))
            pid_write(self._config["pidfile"], pid, action)
            while timeout >= 0:
                try:
                    os.kill(pid, 0)
                except OSError:
                    break
                timeout -= 1
                time.sleep(0.5)
            try:
                os.kill(pid, 0)
            except OSError:
                print("%s (pid %d) does not seem to be running anymore" %
                      (self.prog, pid))
                sys.exit(0)
        pid_quit(self._config["pidfile"], self.prog)
        sys.exit(0)

    def stop_supervisor(self):
        """ Tell the supervisor to stop without touching the children. """
        self.stop("stop_supervisor")

    def stop_children(self):
        """ Tell the supervisor to stop the children. """
        self.send_action("stop_children")

    def wake_up(self):
        """ Tell the supervisor to wake up. """
        self.send_action("wake_up")

    def send_action(self, action):
        """ Tell the supervisor to execute an action. """
        if not self._config.get("pidfile"):
            raise SimplevisorError("action %s requires a pidfile" % action)
        pid = pid_read(self._config["pidfile"])
        if pid:
            print("%s (pid %d) is being told to %s..." %
                  (self.prog, pid, action))
            pid_write(self._config["pidfile"], pid, action)
        elif pid is not None:
            print("%s does not seem to be running anymore" %
                  (self.prog, ))

    def restart(self):
        """ Restart not accepted. """
        raise SimplevisorError(
            "restart command is accepted only with a path")

    def check(self, child=None):
        """
        Confront the expected status with the real status of
        the children aggregating them and return code associated
        with the state.

        0 if everything is fine.

        1 if status is unexpected.
        """
        child = child or self._child
        (child_status, output) = child.check()
        print_nested_list(output, level=0, indent=2)
        if child_status:
            sys.exit(0)
        else:
            sys.exit(1)

    def status(self):
        """ Execute status command. """
        if not self._config.get("pidfile"):
            raise SimplevisorError("status requires a pidfile")
        (status, message) = pid_status(self._config["pidfile"], 60)
        print("%s" % (message, ))
        sys.exit(status)

    def load_status(self):
        """ Load saved status. """
        if self._status_file is not None:
            old_status = self.read_status()
            if old_status is not None:
                self._child.load_status(
                    old_status.get(self._child.get_id(), None))

    def read_status(self):
        """ Read the status from the specified file. """
        try:
            tmp_file = open(self._status_file, "r")
            try:
                status = json.load(tmp_file)
            except ValueError:
                raise SimplevisorError(
                    "Status file not valid: %s" % (self._status_file, ))
            else:
                tmp_file.close()
                return status
        except IOError:
            # self.logger.info("status file not found, continuing.")
            pass

    def save_status(self):
        """ Save the status in the specified file. """
        if self._status_file is None:
            return
        try:
            self.logger.debug("status file: %s", self._status_file)
            status_f = open(self._status_file, "w")
            try:
                status = {self._child.get_id(): self._child.dump_status()}
                json.dump(status, status_f)
                self.logger.debug("status saved: %s", status)
            except StandardError:
                error_type, error, _ = sys.exc_info()
                msg = "error writing status file %s: %s - %s" % \
                      (self._status_file, error_type, error)
                self.logger.error(msg)
                raise SimplevisorError(msg)
            status_f.close()
        except IOError:
            error = sys.exc_info()[1]
            msg = "error writing to status file %s: %s" % \
                  (self._status_file, error)
            self.logger.error(msg)
            raise IOError(msg)

    def sleep_interval(self):
        """ Return the interval between every supervision cycle. """
        value = self._config.get("interval", DEFAULT_INTERVAL)
        result = get_int_or_die(
            value,
            "interval value must be an integer: %s" % (value, ))
        return result

    def pre_run(self):
        """ Before detaching. """
        if isinstance(self._child, service.Service):
            (rcode, _, _) = self._child.start()
            sys.exit(rcode)
        self._child.start()

    def supervise(self):
        """
        Supervise method helper.
        """
        result = {
            'ok': 0,
            'adjusted': 0,
            'failed': 0, }
        t_start = time.time()
        successful = self._child.supervise(result)
        t_end = time.time()
        if successful:
            self.logger.info(
                "supervision cycle executed successfully in %.3fs: "
                "%s services OK, %s services needed adjustment, "
                "%s services failed adjustment",
                t_end - t_start,
                result.get("ok", "unknown"),
                result.get("adjusted", "unknown"),
                result.get("failed", "unknown"))
        else:
            raise SimplevisorError("supervision interrupted/failed")

    def run(self):
        """ Coordinate the job. """
        self.logger.info("started")
        self._running = True
        action = None
        while self._running:
            self.supervise()
            self.save_status()
            wake_time = self.sleep_interval() + time.time()
            self.logger.debug(
                "sleeping for %d seconds", self.sleep_interval())
            while wake_time >= time.time():
                if self._config.get("pidfile"):
                    pid_touch(self._config["pidfile"])
                    action = pid_check(self._config["pidfile"])
                    if action != "":
                        self.logger.info("asked to %s", action)
                        pid_write(self._config["pidfile"], os.getpid())
                    if action in ["quit", "stop_supervisor"]:
                        self._running = False
                    elif action == "stop_children":
                        self._child.stop()
                    elif action == "wake_up":
                        break
                    elif action[:14] == "restart_child ":
                        target = self.get_child(action[14:])
                        target.restart()
                    elif action != "":
                        self.logger.warning("unknown action: %s", action)
                if not self._running:
                    break
                time.sleep(0.5)
        if action != "stop_supervisor":
            self.logger.info("stopping all the children")
            self._child.stop()
        self.logger.info("stopping")
        self.save_status()
        self.logger.info("stopped")
